Sources/AsyncAlgorithms/AsyncBroadcastSequence.swift

```diff
@@ -0,0 +1,246 @@
+      sides[id]?.next(continuation)
+    }
+    
+    static func next(_ state: ManagedCriticalState<State>, id: Int) async -> Result<Element?, Error> {
+      await withUnsafeContinuation { continuation in
+        state.withCriticalRegion { $0.next(id, continuation: continuation) }
+      }
+    }
+    
+    mutating func emit(_ result: Result<Element?, Error>) {
+      for id in sides.keys {
+        sides[id]?.emit(result)
+      }
+    }
+    
+    static func emit(_ state: ManagedCriticalState<State>, result: Result<Element?, Error>) {
+      state.withCriticalRegion { $0.emit(result) }
+    }
+  }
+  
+  struct Iteration {
+    enum Status {
+      case initial(Base)
+      case iterating(Task<Void, Never>)
+      case terminal
+    }
+    
+    var status: Status
+    
+    init(_ base: Base) {
+      status = .initial(base)
+    }
+    
+    static func task(_ state: ManagedCriticalState<State>, base: Base) -> Task<Void, Never> {
+      Task {
+        do {
+          for try await element in base {
+            State.emit(state, result: .success(element))
+          }
+          State.emit(state, result: .success(nil))
+        } catch {
+          State.emit(state, result: .failure(error))
+        }
+      }
+    }
+    
+    mutating func start(_ state: ManagedCriticalState<State>) -> Bool {
+      switch status {
+      case .terminal:
+        return false
+      case .initial(let base):
+        status = .iterating(Iteration.task(state, base: base))
+      default:
+        break
+      }
+      return true
+    }
+    
+    mutating func cancel() {
+      switch status {
+      case .iterating(let task):
+        task.cancel()
+      default:
+        break
+      }
+      status = .terminal
+    }
+    
+    static func start(_ iteration: ManagedCriticalState<Iteration>, state: ManagedCriticalState<State>) -> Bool {
+      iteration.withCriticalRegion { $0.start(state) }
+    }
+    
+    static func cancel(_ iteration: ManagedCriticalState<Iteration>) {
+      iteration.withCriticalRegion { $0.cancel() }
+    }
+  }
+  
+  let state: ManagedCriticalState<State>
+  let iteration: ManagedCriticalState<Iteration>
+  
+  init(_ base: Base) {
+    state = ManagedCriticalState(State())
+    iteration = ManagedCriticalState(Iteration(base))
+  }
+}
+
+
+extension AsyncBroadcastSequence: AsyncSequence {
+  public typealias Element = Base.Element
+  
+  public struct Iterator: AsyncIteratorProtocol {
+    final class Context {
+      let state: ManagedCriticalState<State>
+      var iteration: ManagedCriticalState<Iteration>
+      let id: Int
+      
+      init(_ state: ManagedCriticalState<State>, _ iteration: ManagedCriticalState<Iteration>) {
+        self.state = state
+        self.iteration = iteration
+        self.id = State.establish(state)
+      }
+      
+      deinit {
+        State.cancel(state, id: id)
+        if iteration.isKnownUniquelyReferenced() {
+          Iteration.cancel(iteration)
+        }
```

Comment:
I might be missing something here, but if the last two iterators enter `deinit` at the same time (and both perform the `isKnownUniquelyReferenced` check before releasing `iteration`), could that trigger a race condition such that neither ends up calling `Iteration.cancel`?